### PR TITLE
[WIP] Adding commands for validate and start-upgrade to upgrade script

### PR DESCRIPTION
**Pull request by heschlie**
## Description

Initial go at fleshing out `validate` and `start-upgrade`. Simplistic approach so far, looking for feedback.

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
None required
```

### CODE REVIEW

File: calico_upgrade/pkg/commands/start_upgrade.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+	cfv2 := parsedArgs["--calicov2-config"].(string)
+	cfv3 := parsedArgs["--calicov3-config"].(string)
+
+	clientV3, clientV2, err := upgradeclients.LoadClients(cfv3, cfv2)
+	if err != nil {
+		fmt.Printf("Failed to create Calico API client: %s\n", err)
+		os.Exit(1)
+	}
+
+	_, err = migrate.MigrateData(clientV3, clientV2)
```
Comment (heschlie) [inline calico_upgrade/pkg/commands/start_upgrade.go]: @robbrockbank I wasn't sure if `MigrateData` is supposed to actually write out the data to the datastore and just hasn't been finished or if I'm supposed to implement that here.

Comment (robbrockbank) [reply calico_upgrade/pkg/commands/start_upgrade.go]: it isn't finished yet...

**Comment by tmjd** [comment]
Looks reasonable to me (my 2 :moneybag: - I wanted an emoji for coins/cents)

File: calico_upgrade/pkg/commands/validate.go [hunk 2]
```diff
@@ -52,8 +51,65 @@ Description:
...
+		for _, change := range cData.NameConversions {
+			fmt.Printf(" -  %s: %s -> %s\n", change.Kind, change.Original, change.New)
+		}
+	}
+
+	// After we've converted the names we might end up with clashing names that the user will need to update.
+	if len(cData.NameClashes) > 0 {
+		fmt.Println("The following names clashed after conversion was applied:")
+		for _, nError := range cData.NameClashes {
+			fmt.Println(nError)
```
Comment (bcreane) [inline calico_upgrade/pkg/commands/validate.go]: might be nice to prepend a \<tab\> for each clashing name, or better yet, use the same format you use on line 69: `(" -  %s")`.

File: calico_upgrade/pkg/commands/validate.go [hunk 3]
```diff
@@ -52,8 +51,65 @@ Description:
...
+		}
+	}
+
+	// Errors with data that cannot be converted.
+	if len(cData.ConversionErrors) > 0 {
+		fmt.Println("The following errors were seen during validation, please resolve these errors " +
+			"and run `calico-upgrade validate` again:")
+		for _, cError := range cData.ConversionErrors {
+			fmt.Println(cError)
```
Comment (bcreane) [inline calico_upgrade/pkg/commands/validate.go]: same comment about indent/prepend dash here ...

File: calico_upgrade/pkg/commands/validate.go [hunk 4]
```diff
@@ -52,8 +51,65 @@ Description:
...
+		}
+	}
+
+	// Errors with validation logic, user should report these.
+	if len(cData.ConversedValidationErrors) > 0 {
+		fmt.Println("Errors with validation were seen, please report these to the Calico team " +
+			"on Github by filing an issue (https://github.com/projectcalico/calico/issues):")
+		for _, vError := range cData.ConversedValidationErrors {
+			fmt.Println(vError)
```
Comment (bcreane) [inline calico_upgrade/pkg/commands/validate.go]: and here

File: calico_upgrade/pkg/commands/validate.go (hunk 2 above)
Comment (heschlie) [reply calico_upgrade/pkg/commands/validate.go]: The `String()` method for these has that sort of formatting in it already see https://github.com/projectcalico/calico/blob/calico-upgrade/calico_upgrade/pkg/migrate/migrate.go#L99-L126

Comment (bcreane) [reply calico_upgrade/pkg/commands/validate.go]: ah, cool. thanks for the pointer.

**Pull request merged**